source/ota_over_mqtt_demo.c

```diff
@@ -752,71 +743,152 @@ static void prvProcessIncomingJobMessage( void * pxSubscriptionContext,
 
 /*-----------------------------------------------------------*/
 
-static void prvCommandCallback( void * pCommandContext,
-                                MQTTAgentReturnInfo_t * pReturnInfo )
+static void prvProcessIncomingOtaMessage( void * pvIncomingPublishCallbackContext,
+                                          MQTTPublishInfo_t * pxPublishInfo )
 {
-    TaskHandle_t xTaskToNotify = ( TaskHandle_t ) pCommandContext;
+    bool isMatch = false;
+
+    ( void ) MQTT_MatchTopic( pxPublishInfo->pTopicName,
+                              pxPublishInfo->topicNameLength,
+                              OTA_JOB_ACCEPTED_RESPONSE_TOPIC_FILTER,
+                              OTA_JOB_ACCEPTED_RESPONSE_TOPIC_FILTER_LENGTH,
+                              &isMatch );
 
-    configASSERT( xTaskToNotify );
-    xTaskNotify( xTaskToNotify, pReturnInfo->returnCode, eSetValueWithOverwrite );
+    if( isMatch == true )
+    {
+        prvProcessIncomingJobMessage( pvIncomingPublishCallbackContext, pxPublishInfo );
+    }
+    else
+    {
+        LogWarn( ( "Unhandled OTA message received on topic: %.*s.",
+                   pxPublishInfo->topicNameLength,
+                   pxPublishInfo->pTopicName ) );
+    }
 }
+/*-----------------------------------------------------------*/
+
+static void prvRegisterOTACallback( const char * pTopicFilter,
+                                    uint16_t topicFilterLength )
+{
+    bool isMatch = false;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    uint16_t index = 0U;
+    uint16_t numTopicFilters = sizeof( otaTopicFilterCallbacks ) / sizeof( OtaTopicFilterCallback_t );
+
+
+    bool subscriptionAdded;
+
+    ( void ) mqttStatus;
+
+    /* Match the input topic filter against the wild-card pattern of topics filters
+    * relevant for the OTA Update service to determine the type of topic filter. */
+    for( ; index < numTopicFilters; index++ )
+    {
+        mqttStatus = MQTT_MatchTopic( pTopicFilter,
+                                      topicFilterLength,
+                                      otaTopicFilterCallbacks[ index ].pTopicFilter,
+                                      otaTopicFilterCallbacks[ index ].topicFilterLength,
+                                      &isMatch );
+        assert( mqttStatus == MQTTSuccess );
```

Comment:
I'm really lost here; why are you matching one filter to another?  Why are there two place for filters in this program?

Reply:
This is used in place for comparing a topic filter to a list of demo callbacks to know which callback to register for that topic filter. `MQTT_MatchTopic` is used, instead of parsing the topic in demo, so as to determine the type of topic and the associated callback.

Reply:
Moved to a single handler [vOTAProcessMessage](https://github.com/FreeRTOS/Lab-Project-coreMQTT-Agent/pull/20/commits/1cbafc06422b8955b6398b2b7e6846fc1a0f21d9#diff-dca26c73ee244b83e7c4dbba30ffb673a144a4f937dc8403a25c60c6f1746aafR713)